src-self-hosted/cgen.zig

```diff
@@ -0,0 +1,160 @@
+const Module = @import("Module.zig");
+const ir = @import("ir.zig");
+const Value = @import("value.zig").Value;
+const Type = @import("type.zig").Type;
+const std = @import("std");
+
+const C = link.File.C;
+const Decl = Module.Decl;
+const CStandard = Module.CStandard;
+const mem = std.mem;
+
+/// Maps a name from Zig source to C. This will always give the same output for
+/// any given input.
+fn map(name: []const u8) ![]const u8 {
+    return name;
+}
+
+fn renderType(file: *C, writer: std.ArrayList(u8).Writer, T: Type) !void {
+    if (T.tag() == .usize) {
+        file.need_stddef = true;
+        try writer.writeAll("size_t");
+    } else {
+        switch (T.zigTypeTag()) {
+            .NoReturn => try writer.writeAll("_Noreturn void"),
+            .Void => try writer.writeAll("void"),
+            else => return error.Unimplemented,
+        }
+    }
+}
+
+fn renderFunctionSignature(file: *C, writer: std.ArrayList(u8).Writer, decl: *Decl) !void {
+    const tv = decl.typed_value.most_recent.typed_value;
+    try renderType(file, writer, tv.ty.fnReturnType());
+    const name = try map(mem.spanZ(decl.name));
+    try writer.print(" {}(", .{name});
+    if (tv.ty.fnParamLen() == 0) {
+        try writer.writeAll("void)");
+    } else {
+        return error.Unimplemented;
+    }
+}
+
+pub fn generate(file: *C, decl: *Decl, standard: CStandard) !void {
+    const writer = file.main.writer();
+    const header = file.header.writer();
+    const tv = decl.typed_value.most_recent.typed_value;
+    switch (tv.ty.zigTypeTag()) {
+        .Fn => {
+            try renderFunctionSignature(file, writer, decl);
+            try writer.writeAll(" {");
+
+            const func: *Module.Fn = tv.val.cast(Value.Payload.Function).?.func;
+            const instructions = func.analysis.success.instructions;
+            if (instructions.len > 0) {
+                for (instructions) |inst| {
+                    try writer.writeAll("\n\t");
+                    switch (inst.tag) {
+                        .assembly => {
+                            const as = inst.cast(ir.Inst.Assembly).?.args;
+                            for (as.inputs) |i, index| {
+                                if (i[0] == '{' and i[i.len - 1] == '}') {
+                                    const reg = i[1 .. i.len - 1];
+                                    const arg = as.args[index];
+                                    if (arg.cast(ir.Inst.Constant)) |c| {
+                                        if (c.val.tag() == .int_u64) {
+                                            try writer.writeAll("register ");
+                                            try renderType(file, writer, arg.ty);
+                                            try writer.print(" {}_constant __asm__(\"{}\") = {};\n\t", .{ reg, reg, c.val.toUnsignedInt() });
+                                        } else {
+                                            return error.Unimplemented;
```

Comment:
Can you change these to follow the pattern for unimplemented stuff set by Module.zig and codegen.zig?

Point being that it emits compile errors (pointing to source location) when something is unimplemented.

Reply:
Yeah, for sure